script/core/diagnostics/missing-fields.lua

```diff
@@ -15,61 +15,74 @@ return function (uri, callback)
     guide.eachSourceType(state.ast, 'table', function (src)
         await.delay()
 
+        vm.removeNode(src) -- the node is not updated correctly, reason still unknown
         local defs = vm.getDefs(src)
+        local sortedDefs = {}
         for _, def in ipairs(defs) do
-            if def.type == 'doc.class' and def.bindSource then
-                if guide.isInRange(def.bindSource, src.start) then
+            if def.type == 'doc.class' then
+                if def.bindSource and guide.isInRange(def.bindSource, src.start) then
                     return
                 end
+                local className = def.class[1]
+                if not sortedDefs[className] then
+                    sortedDefs[className] = {}
+                end
+                local samedefs = sortedDefs[className]
+                samedefs[#samedefs+1] = def
             end
             if def.type == 'doc.type.array'
             or def.type == 'doc.type.table' then
                 return
             end
         end
-        local warnings = {}
-        for _, def in ipairs(defs) do
-            if def.type == 'doc.class' then
-                if not def.fields then
-                    return
-                end
 
-                local requiresKeys = {}
-                for _, field in ipairs(def.fields) do
-                    if  not field.optional
-                    and not vm.compileNode(field):isNullable() then
-                        local key = vm.getKeyName(field)
-                        if key and not requiresKeys[key] then
-                            requiresKeys[key] = true
-                            requiresKeys[#requiresKeys+1] = key
-                        end
-                    end
+        local warnings = {}
+        for _, samedefs in pairs(sortedDefs) do
+            local missedKeys = {}
+            local className
+            for _, def in ipairs(samedefs) do
+                className = def.class[1]
+                if not def.fields or #def.fields == 0 then
+                    goto continue
                 end
 
-                if #requiresKeys == 0 then
-                    return
-                end
                 local myKeys = {}
                 for _, field in ipairs(src) do
-                    local key = vm.getKeyName(field)
+                    local key = vm.getKeyName(field) or field.tindex
                     if key then
                         myKeys[key] = true
```

Comment:
The computation of `myKeys` seems independent with the current nested forloop. It only depends on `src`, which is a function argument in the callback.

I suppose this can be **optimized** to execute only once using lazy initialization 😄 
- first define `local myKeys` outside of current nested forloop
- then inside the loop, only compute it `if myKeys == nil`

```lua
local myKeys
for _, samedefs in pairs(sortedDefs) do
  ...
  for _, def in ipairs(samedefs) do
    ...
    -- build myKeys lookup table once
    if myKeys == nil then
      myKeys = {}
      for _, field in ipairs(src) do
        -- same logic
      end
    end
```